augur/filter.py

```diff
@@ -390,10 +461,325 @@ def include_by_query(metadata, include_where):
     >>> include_by_query(metadata, "region=europe")
     {'strain2'}
 
+    If the column referenced in the given query does not exist, skip the filter.
+
+    >>> include_by_query(metadata, "missing_column=value")
+    set()
+
     """
     column, op, value = parse_filter_query(include_where)
-    included = op(metadata[column].astype(str).str.lower(), value.lower())
-    return set(metadata[included].index.values)
+
+    if column in metadata.columns:
+        included_index = op(metadata[column].astype(str).str.lower(), value.lower())
+        included = set(metadata[included_index].index.values)
+    else:
+        # Skip the inclusion filter if the requested column does not exist.
+        included = set()
+
+    return included
+
+
+def construct_filters(args, metadata, sequence_index):
```

Comment:
This is the function that I'm least happy about for a couple of reasons. The main advantage of this function is that it moves a bunch of if/else checks out of `run`, so we can more easily scan that function for the bigger picture operations.

The disadvantage is that this function simply repackages the contents from `run` into a separate place without making it that much simpler. The current interface is difficult to test by doctest or unit test, so it is only tested by functional tests.

I also wonder whether returning lists of dictionaries of the structure `[{"function": <function>, "function_kwargs": {}}]` would be better (clearer) than returning a tuple of function and args.

Reply:
> so we can more easily scan that function for the bigger picture operations.

This alone is important and valuable.

> The current interface is difficult to test by doctest or unit test, so it is only tested by functional tests.

Interface of `augur filter` or `construct_filters(…)`? Can you say more about the impediments you encountered here to writing the kind of tests you'd want?

> …better (clearer) than returning a tuple of function and args.

I think a tuple of function and args is a good representation here. It's what's expected by places it'd be used, like `pd.DataFrame.pipe()` and `functools.partial()`.

Reply:
Alternatively, instead of passing around (func, args) tuples, we could pass around just partially bound functions. This has the nice property of meaning the code just has to handle a list of arbitrary functions, and [partial objects](https://docs.python.org/3/library/functools.html#partial-objects) already have attributes we can read for `--output-log` reporting. We might want to augment the partials with a `.__name__` or custom `.__str__()` routine, but that can happen opportunistically in `--output-log` formatting code.

Reply:
Similarly, we could instead turn the filtering functions you have now into factory functions that take filter params and return a customized filter function. This is akin to `construct_filters()` creating a list of partially-bound functions, but instead the filter functions themselves get delegated that task. This decouples it `construct_filters()` and let's it be customized independently. It also provides a point for each filter factory to verify filter params make sense or raise errors which `construct_filters()` doesn't need to be aware of.

Extending this, I think you could then get rid of `construct_filters()` almost entirely by making the argparse argument types the filter function factories, e.g. `--include-where` is defined with `type=include_by_query` (or whatever [its named](https://github.com/nextstrain/augur/pull/749#discussion_r673558768) :-P) which produces a function based on the `--include-where` value. The "construction" of filters is then delegated, and all that's left is to arrange the various args, e.g. `args.include_where`, into `include_by` and `exclude_by` in the desired order.

Reply:
> Can you say more about the impediments you encountered here to writing the kind of tests you'd want?

I don't like mocking argparse objects for doctests or unit tests (it's a lot of boilerplate setup code that doesn't communicate much about the functions). I also don't like testing a long list of if statements, since it tends to feel like testing `2 + 2 = 4`. Functional tests seem more appropriate here. My main issue with our current functional tests is that I have no sense of what code they cover. I know this isn't everything, but it is important to know when some bit of code is never run by tests.

> ...get rid of `construct_filters()` almost entirely by making the argparse argument types the filter function factories...

Thank you for the great ideas here! This specific approach is most appealing, at least as a place to validate inputs. Given that the current approach makes sense to you though, I'd leave this type of restructuring for a later PR where that's the complete focus of the PR.